### ISSUE TITLE
monomorph regression for `<B:A>` constraint binding

### ISSUE THREAD

**Issue by Antriel** [HaxeFoundation/haxe#9640]
Works for 4.1.2, doesn't with nightly. The main issue is that nightly allows this to compile.

Secondary issue for both 4.1 and nightly is that `Mono` doesn't have `A` bound in the constructor. Not entirely sure if that is an issue or not. But I would expect the first `$type(mono)` to be something like `Mono<{function foo:Void;}>.

```haxe
class Main {
	static function main() {
		var foo:Foo = null;
		var mono = new Mono(foo);
		$type(mono); // `Mono<Unknown<0>>` for both 4.1 and nightly.
		// var foolike:FooLike = mono.bound;
		// Uncommenting ^ would bind it correctly for both 4.1/nightly.
		var barlike:BarLike = mono.bound;
		// ^ 4.1 correctly errors out (not here, but in `new Mono(foo)`). Nightly happily compiles.
		$type(mono); // Both bind to `Mono<BarLike>`. 
	}
}

class Mono<A> {
	public final bound:A;
	public function new<B:A>(obj:B) bound = obj;
}

class Foo { public function foo():Void {} }
typedef FooLike = { function foo():Void; }
typedef BarLike = { function bar():Void; }
```

**Comment by Simn** [comment]
I think I know why this happens, but I'll have to check what is the correct way to fix this.

The problem is that we're constraining one monomorph to another. I deal with this situation when we actually _bind_ two monomorphs together, but if it's a constraint and the constraining monomorph is never bound then there's never any check.

It should be this piece of code in `classify_constraints`:

```ocaml
		let rec check constr = match constr with
			| MMono(m2,name) ->
				begin match m2.tm_type with
				| None ->
					() (* <- here *)
				| Some t ->
					List.iter (fun constr -> check constr) (constraint_of_type name t)
				end;
```

If there are constraints we recursively check them, but in the `None` case we currently don't do anything.

**Comment by Simn** [comment]
> But I would expect the first $type(mono) to be something like `Mono<{function foo:Void;}>`

Why do you expect a structural constraint here? If anything it would be `Mono<Foo>`. I suppose it could also be `Mono<? : Foo>` which would allow the monomorph of `A` to become a ~parent~ child class of `Foo`.

Come to think of it, that might actually be the solution to this.

**Comment by Antriel** [comment]
That would work too. Ultimately I am using `<B:A>` because I want `A` to
unify with children types. But currently in my codebase (temporarily, but
who knows how long that will stay there) I also want it to unify via
structural subtyping.

BTW: I'm not sure where and how atm (away from PC), but the compiler does
generate a structural constraint in some different place where I use
similar pattern.

**Comment by Simn** [comment]
> BTW: I'm not sure where and how atm (away from PC), but the compiler does generate a structural constraint in some different place where I use similar pattern.

That shouldn't be the case because unification when assigning to a class instance type is nominal, not structural. Please let me know in which case (you think) this happens.

**Comment by Simn** [comment]
Also note that `var foolike:FooLike = mono.bound;` won't work with this fix. This was working in 4.1 because constraint checks were delayed to an unspecified point, which lead to the `FooLike` structure being processed before the `Foo` class. With this being reversed now, you would get an `FooLike should be Foo` error.

**Comment by Antriel** [comment]
> That shouldn't be the case because unification when assigning to a class instance type is nominal, not structural. Please let me know in which case (you think) this happens.

Sorry, my bad, it was coming from me handling generating such structure in a macro.

I'm fine with `var foolike:FooLike = mono.bound;` not working, although it does make me wonder whether some constraint to specify that it can be a structure wouldn't be useful. Maybe something like `class Mono<A:{}>`, which now just means it should be an object? Not sure.

In either case the fix in https://github.com/HaxeFoundation/haxe/commit/593041604d2083dc8130b02f9ea8baf86f94269e seems to do the job (and actually better than 4.1, as it forces me to specify a structure explicitly :D).

**Comment by Simn** [comment]
Cool, thanks for checking!

You can already constrain `A:{}`. In your example you could also do `A:FooLike`, which I think is what you're after here.

**Comment by Antriel** [comment]
`A:FooLike` is indeed what I am after. Also tested `A:{}` and it works! I remember testing that when I was creating all this mess (on 4.1) and it didn't work. But testing it now, works both with the fix and 4.1, so I must have been doing it wrong.

The original motivation behind forcing a structure was that I did not have a `FooLike` to actually use.
Thanks for the fast fix. :)

**Comment by Simn** [comment]
Alright, thanks for the very concise example!
